### DISCUSSION TITLE
Linking GNSSNTRIPClient and RTCMReader using in-memory stream with BytesIO

### DISCUSSION THREAD

**Discussion by dgashby** [semuconsulting/pygnssutils#106] [Q&A] [answered]
Hi,
I'm trying to create an in-memory stream to pass messages on to an RTCMReader using the following:
```
ntrip_rtcm_stream = io.BytesIO(b"")

rtr = RTCMReader(ntrip_rtcm_stream)

gnc = GNSSNTRIPClient()
gnc.run(
    server=NTRIP_SERVER,
    port=NTRIP_PORT,
    https=0,
    mountpoint=NTRIP_MOUNTPOINT,
    datatype="RTCM",
    ntripuser=NTRIP_USER,
    ntrippassword=NTRIP_PASSWORD,
    ggainterval=-1,
    output=ntrip_rtcm_stream,
    verbosity=2,
)
    while True:
        try:                                                                                                                       
            raw, msg = rtr.read()
	    if msg is not None:
                log.info("RTCM:" + msg)

```

I'm seeing the verbose logging from within the GNSSNTRIPClient to indicate various message types are being received from the NTRIP server, however rtr.read() doesn't seem to be reading anything.  When I strip out the RTCMReader and just try to read the ntrip_rtcm_stream it doesn't return any bytes either.  If I replace the BytesIO stream with an actual file, it does have RTCM data written to it.

Any ideas why the in-memory stream isn't working for me?

Thanks,

Dean

**Comment by semuadmin** [comment, marked as answer]
Hi @dgashby, 

Couple of issues with the code...

1. Not sure `BytesIO` can be used in quite this way e.g. consider the following:
```python
from io import BytesIO

b = BytesIO(b"1234")
print(b.getvalue())  # outputs b'1234'
print(b.read(2))  # outputs b'12'
print(b.read(2))  # outputs b'34'
print(b.read())  # outputs b''

b.write(b"5678")
print(b.getvalue())  # outputs b'12345678'
print(b.read())  # still outputs b''

```

2. As per the [API documentation](https://www.semuconsulting.com/pygnssutils/pygnssutils.html#pygnssutils.gnssntripclient.GNSSNTRIPClient.run), the GNSSNTRIPClient `output` argument supports the following specific writeable output media classes (*intended to cover the most common use cases*):

<img width="559" alt="gnssntripclient_api" src="https://github.com/user-attachments/assets/a16f52a8-b923-4493-b40b-aaee1298375a">

    - BufferedWriter e.g. binary file, for raw data bytes
    - TextIOWrapper e.g. text file, for parsed data as a string
    - Serial e.g. a GNSS receiver's UART or USB port
    - Socket e.g. to broadcast to a local TCP/IP network
    - Queue e.g. for asynchronous processing in a separate thread

For a more flexible approach, I recommend you take a look at the [ntrip_client_generic.py](https://github.com/semuconsulting/pygnssutils/blob/main/examples/ntrip_client_generic.py) example in the [\examples](https://github.com/semuconsulting/pygnssutils/tree/main/examples) folder. You can substitute the simple `print()` statement in the `datathread` with whatever you want to do with the raw or parsed RTCM data you receive.

Hope this helps.

**Reply by dgashby** [reply]
Thank you for the quick response.  I had not noticed ntrip_client_generic.py, it looks like using a Queue combined with a separate thread to clear the queue will solve my problem.  Very much appreciate your help.
